### PR TITLE
feat: added plugins at shared ES/OS clients

### PR DESCRIPTION
**Pull request by igpetrov**
## Description

feat: added plugins at shared ES client

Similar to this: https://github.com/camunda/camunda/issues/21559

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

Related to: https://github.com/camunda/camunda/issues/21559

### CODE REVIEW

**Comment by igpetrov** [comment]
Hi @romansmirnov 

Similar to https://github.com/camunda/camunda/issues/21559, I implemented plugin support for shared ES/OS clients. I know that they are not currently used, but I did it anyway so that in future we don't forget about this feature.

Please take a look and let me know what you think.

File: search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java [hunk 1]
```diff
@@ -45,24 +48,44 @@ public final class OpensearchConnector {
...
+        new JacksonConfiguration(configuration).createObjectMapper(),
+        DefaultCredentialsProvider.create(),
+        new PluginRepository());
   }
 
   public OpensearchConnector(
-      final ConnectConfiguration configuration, final ObjectMapper objectMapper) {
+      final ConnectConfiguration configuration,
+      final ObjectMapper objectMapper,
+      final AwsCredentialsProvider credentialsProvider,
```
Comment (romansmirnov) [inline search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java:66]: ❓ Why was this change necessary?

**Review by romansmirnov — COMMENTED** [review]
In general, it looks good, pretty straightforward. However, I have a question because I don't understand why the `OpensearchConnector` now also accepts a `AwsCredentialsProvider`.

File: search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java (hunk 1 above)
Comment (igpetrov) [reply search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java:66]: It was very hard to test/mock: because method `boolean shouldCreateAWSBasedTransport()` explicitly invokes `DefaultCredentialsProvider.create()`. See, the  `DefaultCredentialsProvider` initializes in chain `EnvironmentVariableCredentialsProvider`, `SystemPropertyCredentialsProvider`, and few others. This means, that people (like myself), who have a set of configs for AWS accounts as env vars, will have AWS initialized, thus making impossible to test local instance of OpenSearch.

So I decided that since there are multiple implementations of the `AwsCredentialsProvider`, it might not hurt to move it into constructor. Please keep in mind, that the backwards compatibility is respected, and when not defined, setting `DefaultCredentialsProvider.create()`.

WDYT?

**Comment by igpetrov** [comment]
Adding @npepinpe since @romansmirnov seems to be on FTO

**Comment by igpetrov** [comment]
Also adding @mustafadagher who has some context on this topic

File: search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java [hunk 2]
```diff
@@ -45,24 +48,44 @@ public final class OpensearchConnector {
 
   private final ConnectConfiguration configuration;
   private final ObjectMapper objectMapper;
+  private final PluginRepository pluginRepository;
+
+  private AwsCredentialsProvider credentialsProvider;
```
Comment (mustafadagher) [inline search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java:53]: 🔧 this could be final, since it's either passed to or initialized in a constructor.

File: search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java [hunk 3]
```diff
@@ -125,14 +152,21 @@ private HttpHost getHttpHost(final ConnectConfiguration osConfig) {
...
+    } catch (final URISyntaxException e) {
       throw new SearchClientConnectException("Error in url: " + osConfig.getUrl(), e);
     }
   }
 
   protected HttpAsyncClientBuilder configureHttpClient(
-      final HttpAsyncClientBuilder httpAsyncClientBuilder, final ConnectConfiguration osConfig) {
+      final HttpAsyncClientBuilder httpAsyncClientBuilder,
+      final ConnectConfiguration osConfig,
+      final HttpRequestInterceptor... interceptors) {
```
Comment (mustafadagher) [inline search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java:163]: ❓ why an array of interceptors here, when we only ever pass one?

**Review by mustafadagher — APPROVED** [review]
Thanks @igpetrov! LGTM 🚀 

I've only left a few questions/suggestions in comments below.

File: search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java (hunk 2 above)
Comment (igpetrov) [reply search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java:53]: Good point. Will make it!

File: search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java (hunk 3 above)
Comment (igpetrov) [reply search/search-client-connect/src/main/java/io/camunda/search/connect/os/OpensearchConnector.java:163]: There was a case when configuration required multiple plugin repositories / interceptors, so we decided that we will support an array of interceptors. In addition to that, the plugin functionality will grow in similar way as Optimize C7 had.

File: search/search-client-connect/src/test/java/io/camunda/search/connect/os/OpensearchConnectorTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.mockito.Mockito;
+import org.opensearch.client.transport.httpclient5.ApacheHttpClient5Transport;
+
+class OpensearchConnectorTest {
+
+  @RegisterExtension
+  // Native OS clients refuse to work while
+  // actual HTTP(S) connection is available
```
[the remark this answers is not in the document]
Comment (igpetrov) [reply search/search-client-connect/src/test/java/io/camunda/search/connect/os/OpensearchConnectorTest.java:38]: So, if I try executing any non-existing HTTP endpoint with `org.opensearch.client.transport.httpclient5.ApacheHttpClient5Transport` (e.g., `localhost:1234`, assuming it is unavailable), the HTTP client won't even do anything, just silently fail. Therefore in this case, I have to actually kick on a wiremock. And then it does perform an HTTP call, including adding headers.

This is different to ES client, where header will be added to whatever non-existing endpoint: see similar `ElasticsearchConnectorTest`.
